oxidation/libparsec/crates/types/src/certif.rs

```diff
@@ -16,81 +16,53 @@ use crate::{
-                let compressed = author_verify_key.verify(signed)?;
-                let mut serialized = vec![];
-                ZlibDecoder::new(&compressed[..])
-                    .read_to_end(&mut serialized)
-                    .map_err(|_| DataError::Compression)?;
-                let obj: $name =
-                    ::rmp_serde::from_slice(&serialized).map_err(|_| DataError::Serialization)?;
-                match (&obj.author, expected_author) {
-                    (CertificateSignerOwned::User(ref a_id), CertificateSignerRef::User(ea_id)) => {
-                        if a_id == ea_id {
-                            Ok(obj)
-                        } else {
-                            Err(DataError::UnexpectedAuthor {
-                                expected: ea_id.clone(),
-                                got: Some(a_id.clone()),
-                            })
-                        }
-                    }
-                    (_, CertificateSignerRef::Root) => Ok(obj),
-                    (_, CertificateSignerRef::User(ea_id)) => Err(DataError::UnexpectedAuthor {
-                        expected: ea_id.clone(),
-                        got: None,
-                    }),
-                }
-            }
-        }
-    };
+fn verify_and_load<T>(signed: &[u8], author_verify_key: &VerifyKey) -> DataResult<T>
+where
+    T: for<'a> Deserialize<'a>,
+{
+    let compressed = author_verify_key.verify(signed)?;
+    let mut serialized = vec![];
+    ZlibDecoder::new(&compressed[..])
+        .read_to_end(&mut serialized)
+        .map_err(|_| DataError::Compression)?;
+    rmp_serde::from_slice(&serialized).map_err(|_| DataError::Serialization)
 }
 
-macro_rules! impl_verify_and_load_no_root {
-    ($name:ident) => {
-        impl $name {
-            pub fn verify_and_load(
-                signed: &[u8],
-                author_verify_key: &VerifyKey,
-                expected_author: &DeviceID,
-            ) -> DataResult<$name> {
-                let compressed = author_verify_key.verify(signed)?;
-                let mut serialized = vec![];
-                ZlibDecoder::new(&compressed[..])
-                    .read_to_end(&mut serialized)
-                    .map_err(|_| DataError::Compression)?;
-                let obj: $name =
-                    ::rmp_serde::from_slice(&serialized).map_err(|_| DataError::Serialization)?;
-                if &obj.author != expected_author {
-                    Err(DataError::UnexpectedAuthor {
-                        expected: expected_author.clone(),
-                        got: Some(obj.author.clone()),
-                    })
-                } else {
-                    Ok(obj)
-                }
-            }
+fn check_author_allow_root(
+    author: &CertificateSignerOwned,
+    expected_author: CertificateSignerRef,
+) -> DataResult<()> {
+    match (author, expected_author) {
+        (CertificateSignerOwned::User(a_id), CertificateSignerRef::User(ea_id))
```

Comment:
Could you use more distinct naming for the variables `a_id` & `ea_id` ?

Reply:
Done